### DISCUSSION TITLE
Going modular

### DISCUSSION THREAD

**Discussion by urob** [urob/zmk-config#73] [Announcements]
Currently, this config relies on a custom ZMK fork for the building. Historically, that fork included many experimental features, many of which are now part of upstream ZMK or are available as ZMK modules.

My hope is that with the module ecosystem growing, I'll be eventually able to build against upstream ZMK and a few modules. In c8b436964ea0c25dda10aba9adbdb93fdacd13ab, I started making this transition. ~Instead of using my [`main`](https://github.com/urob/zmk/tree/main) ZMK branch to build my firmware, I'm now using [`going-modular`](https://github.com/urob/zmk/tree/going-modular)~. 

**Update:** The modular setup is now merged into the [`main`](https://github.com/urob/zmk/tree/main) branch of my ZMK repo. The old `main` has been renamed to `main-3.5`. Neither `main-3.5` nor `going-modular` will be maintained going forward.

The following tracks the remaining patches merged in `main`.

- [x] [`require-prior-idle`](https://github.com/zmkfirmware/zmk/pull/1387) (merged upstream)
- [x] [combo + hold-tap fixes](https://github.com/zmkfirmware/zmk/pull/1411) (merged upstream)
- [x] [`masked-mods` fix](https://github.com/zmkfirmware/zmk/pull/1412) (merged upstream)
- [x] [`hold-on-release` for `balanced`](https://github.com/zmkfirmware/zmk/pull/1423) (merged upstream)
- [x] [rolling `repeat` fix](https://github.com/zmkfirmware/zmk/pull/1828) (merged upstream)
- [x]  [Pointer core/mouse keys](https://github.com/zmkfirmware/zmk/pull/1991) (merged upstream)
- [x] ~[Pointer/movement scrolling PR](https://github.com/zmkfirmware/zmk/pull/2027)~ (replaced by new mouse PR)
- [x] [New pointer/movement scrolling PR](https://github.com/zmkfirmware/zmk/pull/2477) (merged upstream)
- [x] [`tri-state` behavior PR](https://github.com/zmkfirmware/zmk/pull/1366) (replaced by [`tri-state` module](https://github.com/urob/zmk-tri-state))
- [x] [`smart-word` behavior PR](https://github.com/zmkfirmware/zmk/pull/1451) (replaced by [`auto-layer` module](https://github.com/urob/zmk-auto-layer))
- [x]  [`hold-on-release` for `tap-preferred`](https://github.com/zmkfirmware/zmk/commit/430e7b606199394b747b99c479ee2764d65425c1) (no replacement needed [^1])
- [x] ~[Zen display & battery tweaks](https://github.com/caksoylar/zmk/tree/caksoylar/zen-v1%2Bv2)~ (removed from my fork as my Zen display is almost dead anyhow)
- [x] [Leader key PR](https://github.com/zmkfirmware/zmk/pull/1380) (replaced by [`leader-key` module](https://github.com/urob/zmk-leader-key))
- [x] ~[Generic desktop usage page](https://github.com/zmkfirmware/zmk/pull/2473)~ (removed from my fork as it didn't work via BLE for me)
- [x] [Fix child node dependencies](https://github.com/zmkfirmware/zephyr/pull/40) (I am now overloading `zephyr` in my west manifest directly)
- [x] [Fix `test.py`](https://github.com/zmkfirmware/zmk/pull/2425) (merged upstream)
- [x] ~[Workaround](https://github.com/zmkfirmware/zmk/commit/3ad41f76bea89aff3d8110e21ca261a5123ca501) for an upstream `west` issue ([#725](https://github.com/zephyrproject-rtos/west/issues/725))~ (reorganized my workspace to avoid the issue)

**Update:** Done - We are on stock ZMK 🥳🥳🥳

[^1]: I kept this commit around as a courtesy for people who wanted to play around with hold-tap flavors and still use positional hold-taps. But since I am only using the balanced flavor myself, there's no need to keep it in my ZMK branch.

**Comment by dhruvinsh** [comment]
Glad to see that `tri-state` module proof of concept working out great. :)

Awesome work. 👍🏼

**Reply by urob** [reply]
Yes, thanks for figuring out how all the modularization stuff! I also used it as a reference for modularizing the updated `auto-layers` behavior. Really very useful!

**Reply by urob** [reply]
Btw, not sure if you want to keep around the old `num-word` implementation. But maybe it's worth adding a comment about how it is different from the new `auto-layers` so people don't get confused?

**Reply by dhruvinsh** [reply]
> Btw, not sure if you want to keep around the old num-word implementation

definitely I don't plan to develop it, it was just your work with proving that modularization can be done. I will have note on the repo pointing to your valid work, In-fact I was just waiting for that :)

I will keep the `swapper` aka `zmk-tri-state` as it is as Nick haven't posted any changes yet, let see, 🤞

**Comment by infused-kim** [comment]
Amazing work. Thank you for sharing it all!

**Comment by minusfive** [comment]
Thanks @urob! Updated my config, no issues, working flawlessly.

**Comment by urob** [comment]
Update: as of d1dfb03d8e566def0645bb0b8f629a11b7a46b71 I am running vanilla ZMK! 🎉

Technically, I am still linking to my remote. But this is only so I can pin the current state of upstream `main` since there's isn't yet a release with the merged pointer PR. There are no custom patches left in my `v0.1+mouse` branch.
